DSNP on Frequency Publishing (#204)
LibertyDSNP/spec: pages/Frequency/Publishing.md

| 6 | [Update](../DSNP/Types/Update.md) | YES | TBD | TBD | [`Parquet`](https://libertydsnp.github.io/frequency/common_primitives/schema/enum.ModelType.html#variant.Parquet) | [`IPFS`](https://libertydsnp.github.io/frequency/common_primitives/schema/enum.PayloadLocation.html#variant.IPFS) |

Source code for each schema is also located in the [LibertyDSNP/schemas](https://github.com/LibertyDSNP/schemas) repository.

## DSNP Non-Batched Announcements

Frequency uses an on-chain data structure for storage of `GraphChange` Announcements.
Each individual announcement is submitted using the [`add_onchain_message`](https://libertydsnp.github.io/frequency/pallet_messages/pallet/enum.Call.html#variant.add_onchain_message) extrinsic.

## DSNP Batched Announcements

Frequency uses [DSNP Batch Publications](../DSNP/BatchPublications.md) for Announcements that are batched.
The parquet file is stored on IPFS, but it is discovered through the Frequency Message.

> Frequency Messages maintain the data of when the data was published and what Providers MSA published it.

Reviewer: > Frequency Messages maintain the data of when the data was published...

Not sure what you mean here.

Author: @shannonwells How does this work?
```suggestion
Frequency Messages maintain the metadata of where and when the Batch Publication was published and which Provider MSA published it.
```

> It is the publisher's responsibility to maintain the IPFS pin so that the batch file is continuously available.

Reviewer: Will the intended audience know what an IPFS pin is?  Maybe this should link to an IPFS reference here or at the bottom.

DSNP Batch Publications [MUST be validated](./Validation.md) upon fetching to ensure data and permission integrity.

## Ordering

Frequency Messages are well ordered within a Schema

1. Frequency: Block number ascending
2. Frequency: Transaction index ascending
3. DSNP Standard: Order Announcements in a Batch Publication File by row appearance order

### Ordering Across Schemas

> Frequency does not provide complete order information for Messages across Schemas.

Reviewer: ```suggestion
Frequency does not provide complete ordering metadata for Messages across Schemas.
```

Reviewer: Maybe "ordering" or "ordering information"....

> Frequency does not provide complete order information for Messages across Schemas.

Reviewer 2: We do provide order across different schemas in the same block. `Index` stored on each Message is a total index in that block across all schemas.
